packages/babel-parser/src/parser/comments.js

```diff
@@ -284,4 +290,104 @@ export default class CommentsParser extends BaseParser {
 
     stack.push(node);
   }
+
+  processComment(node: Node): void {
+    const { unattachedCommentStack } = this.state;
+    const commentStackLength = unattachedCommentStack.length;
+    if (commentStackLength === 0) return;
+    let i = commentStackLength - 1;
+    const lastComments = unattachedCommentStack[i];
+
+    if (lastComments.start === node.end) {
+      lastComments.leadingNode = node;
+      i--;
+    }
+    // invariant: for all 0 <= j <= i, c = commentStack[j] satisfies c.end < node.end
+    const nodeStart = node.start;
+    for (; i >= 0; i--) {
+      const commentWS = unattachedCommentStack[i];
+      const commentEnd = commentWS.end;
+      if (commentEnd > nodeStart) {
+        // by definition of commentWhiteSpace, this implies commentWS.start > nodeStart
+        // so node can be a containerNode candidate
+        const { containerNode } = commentWS;
+        if (containerNode === null) {
+          commentWS.containerNode = node;
+        } else {
+          if (containerNode.end - containerNode.start < node.end - nodeStart) {
+            unattachedCommentStack.splice(i, 1);
+          } else if (node.type !== "Program") {
+            // we have a node share the same length of containerNode, but its finishNode is invoked later
+            // than containerNode, so this node is the outer node. E.g. ExpressionStatement contains a VariableDeclaration
+            commentWS.containerNode = node;
+          }
+        }
+      } else if (commentEnd === nodeStart) {
+        commentWS.trailingNode = node;
+        break;
+      }
+    }
+  }
+
+  finalizeComment() {
+    const { commentStack } = this.state;
+    const commentStackLength = commentStack.length;
+    for (let i = 0; i < commentStackLength; i++) {
+      const commentWS = commentStack[i];
+      if (commentWS.leadingNode !== null || commentWS.trailingNode !== null) {
+        if (commentWS.leadingNode !== null) {
+          commentWS.leadingNode.trailingComments = commentWS.comments;
+        }
+        if (commentWS.trailingNode !== null) {
+          commentWS.trailingNode.leadingComments = commentWS.comments;
+        }
+      } else {
+        const {
+          containerNode: node,
+          comments,
+          start: commentStart,
+        } = commentWS;
+        if (this.input.charCodeAt(commentStart - 1) === charCodes.comma) {
+          switch (node.type) {
+            case "ObjectExpression":
+            case "ObjectPattern":
+              this.adjustInnerComments(node, node.properties, comments);
+              break;
+            case "CallExpression":
```

Comment:
We also need to do this for `OptionalCallExpression`, and maybe for function definitions?

Reply:
Yes it is addressed in https://github.com/babel/babel/pull/13521/commits/e5f09a5a82cb3b31906a4168a4eff16815fec604.

Just in case, are you reviewing on old diff?

Reply:
Uh I think I was accidentally reviewing a specific commit :sweat_smile: